### PR TITLE
Implement strict flag

### LINKED ISSUE #127: Incorrect names in SVDs
Many Fujitsu and Freescale SVDs have incorrect enumerated value names that start with digit like:
```
                <enumeratedValue>
                  <name>0</name>
                  <description>Configured interrupt is not detected.</description>
                  <value>#0</value>
                </enumeratedValue>
```
For now `svd2rust` just adds `_` prefix.
But `check_name` blocks this:
https://github.com/rust-embedded/svd/blob/1d0833cd0ebe8f527767bc70709b684d6195f418/src/svd/enumeratedvalue.rs#L90

What should we do:
1. Remove this validation check.
2. Add `_` prefix during parsing.
3. Ignore incorrect files.
4. Allow digit on start of name.

cc @therealprof  @Emilgardis

### PR DESCRIPTION
**Pull request by brainstorm**
Fixes #127 among others

### CODE REVIEW

**Comment by burrbull** [comment]
```rust
            #[cfg(feature="strict")]
            if fields.is_empty() {
                return Err(SVDError::EmptyFields)?;
            }
```
Other `Empty*`s and `NameError` should also be disabled by default.

Rebase on master.

File: src/svd/writeconstraint.rs [hunk 1]
```diff
@@ -45,6 +45,7 @@ impl Parse for WriteConstraint {
                 _ => Err(SVDError::UnknownWriteConstraint(tree.clone()).into()),
             }
         } else {
+            //#[cfg(feature="strict")]
```
Comment (burrbull) [inline src/svd/writeconstraint.rs:48]: ?

**Comment by brainstorm** [comment]
@burrbull Would you relax any of the following checks on SVDs? Just a subset? None?:

https://github.com/rust-embedded/svd2rust/blob/3b2e8a535f1e4e6b1db7cd1ff05f6fb7f0c89d03/ci/script.sh#L82

**Comment by burrbull** [comment]
> @burrbull Would you relax any of the following checks on SVDs? Just a subset? None?:
> 
> https://github.com/rust-embedded/svd2rust/blob/3b2e8a535f1e4e6b1db7cd1ff05f6fb7f0c89d03/ci/script.sh#L82

Yes, in `svd-parser` we can relax this check, I think.
Saying about `svd2rust` you need use `write_with_zero` (which shall be unsafe, bug) instead of `write`.

Also we should relax `check_name*`, as I said earlier.

**Comment by brainstorm** [comment]
> Saying about svd2rust you need use write_with_zero (which shall be unsafe, bug) instead of write.

Sorry, but I don't understand what you mean here :-S

**Comment by burrbull** [comment]
> > Saying about svd2rust you need use write_with_zero (which shall be unsafe, bug) instead of write.
> 
> Sorry, but I don't understand what you mean here :-S

`write` function of register in `svd2rust` expects that register has reset value for safe writing. As incorrect value can break controller.
But also we have `write_with_zero` method that doesn't have this restriction.

cc @therealprof

**Comment by Emilgardis** [comment]
I'd prefer if this was on runtime and not compiletime. That can be done with either "unstrict" functions, a bool flag to functions or with recoverable errors in `Ok` path.

**Comment by brainstorm** [comment]
@Emilgardis Can you elaborate a bit more? Not sure what you mean with what you propose :-S

**Comment by burrbull** [comment]
> I'd prefer if this was on runtime and not compiletime. That can be done with either "unstrict" functions, a bool flag to `functions` or with recoverable errors in `Ok` path.
> 
> My preference is the latter.
> 
> Something like this
> 
> ```rust
> pub(crate) fn check_name(name: &str, tag: &str) -> Result<Option<E>> {
> ...
> }
> ```
> 
> Or similar

I see 2 ways to implement your suggestion:
- use global state (statics) to control current configuration
- rewrite all parsing system with recoverable errors (too much complication in my view)

**Comment by brainstorm** [comment]
> I see 2 ways to implement your suggestion:

Since it doesn't look easy (to my inexperienced eyes), wouldn't it make sense to reserve those changes for a separate PR?

**Review by Emilgardis — COMMENTED** [review]
I'm fine with merging when clippy is satisfied, working on [another approach](https://github.com/Emilgardis/svd/commit/6f0eace571eee42098073ad90d14cccc89b7174a) however based on what this touches.

I'd love to retrieve a sample SVD that has this behaviour fwiw.

also, could you rebase instead of doing a merge?

Easiest way to do that is probably

`git pull upstream master --no-commit`

**Comment by brainstorm** [comment]
> also, could you rebase instead of doing a merge?
> 
> Easiest way to do that is probably
> 
> `git pull upstream master --no-commit`

That doesn't seem to work. OTOH, GitHub allows you to squash this PR as you merge it with one clicks, could you do that instead, @Emilgardis?

**Comment by therealprof** [comment]
> `git pull upstream master --no-commit`

`git pull` can be a very good footgun. I usually recommend not using it at all and if one "must", to change the pull strategy to rebase. Just get used to
```
git fetch --all
git rebase --autostash
```

or even add an alias for it.

> That doesn't seem to work. OTOH, GitHub allows you to squash this PR as you merge it with one single click. Could you do that instead, @Emilgardis?

That doesn't work for a number of reasons. Would you please do a `git rebase master` followed by a force push?

**Comment by brainstorm** [comment]
> > `git pull upstream master --no-commit`
> 
> `git pull` can be a very good footgun. I usually recommend not using it at all and if one "must", to change the pull strategy to rebase. Just get used to
> 
> ```
> git fetch --all
> git rebase --autostash
> ```
> 
> or even add an alias for it.
> 
> > That doesn't seem to work. OTOH, GitHub allows you to squash this PR as you merge it with one single click. Could you do that instead, @Emilgardis?
> 
> That doesn't work for a number of reasons. Would you please do a `git rebase master` followed by a force push?

I'm doing precisely that but a conflict on src/svd/clusterinfo.rs ruins the 1-commit rebase nirvana:

```
$ git rebase master
First, rewinding head to replay your work on top of it...
Applying: Implement strict feature flag for SVD parsing.
Applying: ci: fix build step
Using index info to reconstruct a base tree...
M	.github/workflows/ci.yml
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
Applying: Fix clippy for MSRV 1.36
Using index info to reconstruct a base tree...
M	.github/workflows/clippy.yml
M	src/error.rs
M	src/svd/clusterinfo.rs
M	src/svd/device.rs
M	src/svd/enumeratedvalue.rs
M	src/svd/enumeratedvalues.rs
M	src/svd/field.rs
M	src/svd/fieldinfo.rs
M	src/svd/interrupt.rs
M	src/svd/peripheral.rs
M	src/svd/registerinfo.rs
M	tests/bad_svd.rs
Falling back to patching base and 3-way merge...
Auto-merging src/svd/registerinfo.rs
Auto-merging src/svd/peripheral.rs
Auto-merging src/svd/clusterinfo.rs
CONFLICT (content): Merge conflict in src/svd/clusterinfo.rs
Auto-merging src/error.rs
error: Failed to merge in the changes.
Patch failed at 0003 Fix clippy for MSRV 1.36
hint: Use 'git am --show-current-patch' to see the failed patch
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
```

Because I need to address that conflict, then `git commit -am` it and git force push, so 2 commits at the end :-S

**Comment by therealprof** [comment]
I've fixed this problem for you.

**Comment by brainstorm** [comment]
@therealprof Thanks, just did 15 seconds after you, applying `git config --global rerere.enabled 1`, new flag I didn't know it existed :D

File: src/svd/registerinfo.rs [hunk 2]
```diff
@@ -330,6 +331,7 @@ impl Encode for RegisterInfo {
 }
 
 #[cfg(test)]
+
```
Comment (Emilgardis) [inline src/svd/registerinfo.rs:334]: This is not needed

**Review by Emilgardis — APPROVED** [review]
bors r+

**Comment by brainstorm** [comment]
Thanks for merging folks!

@Emilgardis:

> I'd love to retrieve a sample SVD that has this behaviour fwiw.

Which behavior? FWIW, there's a bunch of "broken" SVDs [referenced on the svd2rust CI script](https://github.com/rust-embedded/svd2rust/blob/3b2e8a535f1e4e6b1db7cd1ff05f6fb7f0c89d03/ci/script.sh#L82), I thought it could be useful for [your current approach/branch](https://github.com/Emilgardis/svd/commit/6f0eace571eee42098073ad90d14cccc89b7174a).

**Pull request merged**